Add failing test for async end

dignifiedquire opened the pull request:
So I'm trying to adhere to an interface based on node.js streams by doing the work in a pull-stream and then calling`pullStreamToStream` on it. The interface test writes to the stream and then ends it, but because the internal processing is asynchronous the stream is ended before all the results are emitted. 

I've a failing test case here and would love some feedback if this is actually a bug to be fixed here or if this should be solved in a different way.

dignifiedquire committed:
add failing test for async end

dominictarr commented:
writing a node stream as a pull-stream is the intended use of this library, but there are a lot a undisiplined ways that you can use node streams, and it may be difficult to support all of them.
It sounds like this one should be reasonable though

and also the source stream is incorrect. `CS` should take `(sink, source)` but you've given it `(through, sink)` which doesn't actually make sense.

it does look like the code I wrote is intended to do something like you describe
end: end the source side
https://github.com/pull-stream/pull-stream-to-stream/blob/master/index.js#L39-L44
destroy: end both sides
https://github.com/pull-stream/pull-stream-to-stream/blob/master/index.js#L110-L116

but it also appears very out of date, I stopped using `addPipe` ages ago, so obviously I havn't touched this in a while.

dominictarr commented:
what kind of stream are you wanting to convert? is it a source, sink, through, or duplex?

dignifiedquire committed:
fix source/through confusion

dignifiedquire commented:
> and also the source stream is incorrect. CS should take (sink, source) but you've given it (through, sink) which doesn't actually make sense.

fixed (I hope)

> is it a source, sink, through, or duplex?

The original code is I think a through, ref: https://github.com/ipfs/js-ipfs/blob/pull/src/core/ipfs/files.js#L17-L21.

dignifiedquire committed:
Better handling of through streams

dignifiedquire committed:
emit drain after back pressure was exerted

dignifiedquire commented:
@dominictarr so it seems through streams weren't handled very well, which I fixed now. I also added a fix to emit `drain` when back pressure was exerted (`write` returned `false`), so that a piped stream would resume again.

dominictarr commented:
thanks! I've merged this into 1.3.1, and also I updated the tests to use pull-stream@3 style.
also, I'll add you as a collaborator & npm owner to this module, thanks for the great work!

dignifiedquire commented:
I und think your forgot to push to github, the new releasr doesn't seem to be on on master.

dominictarr commented:
sorry, I was on the wrong branch. fixed now!
